addons/account/models/account_analytic_account.py

```diff
@@ -18,26 +19,40 @@ class AccountAnalyticAccount(models.Model):
     @api.depends('line_ids')
     def _compute_invoice_count(self):
         sale_types = self.env['account.move'].get_sale_types(include_receipts=True)
-        domain = [
-            ('move_line_id.move_id.move_type', 'in', sale_types),
-            ('account_id', 'in', self.ids)
-        ]
-        groups = self.env['account.analytic.line']._read_group(domain, ['move_line_id.move_id:count_distinct'], ['account_id'])
-        moves_count_mapping = dict((g['account_id'][0], g['account_id_count']) for g in groups)
+
+        query = self.env['account.move.line']._search([
+            ('move_id.state', '=', 'posted'),
+            ('move_id.move_type', 'in', sale_types),
+        ])
+        query.add_where('account_move_line.analytic_distribution ?| array[%s]', [str(account_id) for account_id in self.ids])
+        mapped_data = defaultdict(set)
+
+        query_string, query_param = query.select('analytic_distribution', 'move_id')
+        self._cr.execute(query_string, query_param)
+        for data in self._cr.dictfetchall():
+            for analytic_account_id in data['analytic_distribution']:
+                mapped_data[int(analytic_account_id)].add(data['move_id'])
         for account in self:
-            account.invoice_count = moves_count_mapping.get(account.id, 0)
+            account.invoice_count = len(mapped_data.get(account.id, []))
```

Comment:
This seems very complicated for what it does, maybe I'm wrong, or (probably) William will disagree with me, but I think that using a simple SQL query here would make it simpler.
I'm not sure about performances though.

```python
        sale_types = self.env['account.move'].get_sale_types(include_receipts=True)
        self._cr.execute("""
            SELECT jsonb_object_keys(move_line.analytic_distribution) AS analytic_account_id, count(DISTINCT(move.id)) AS invoice_amount
            FROM account_move_line move_line
            JOIN account_move move ON move_line.move_id = move.id
            WHERE move.state = 'posted' AND move.move_type IN %s
            AND move_line.analytic_distribution ?| %s
            GROUP BY jsonb_object_keys(move_line.analytic_distribution)
        """, [tuple(sale_types), [str(account_id) for account_id in self.ids]])
        mapped_data = {int(account_data['analytic_account_id']): account_data['invoice_amount'] for account_data in self._cr.dictfetchall()}
        for account in self:
            account.invoice_count = mapped_data.get(account.id, [])
```
To me, this is easier to read and requires less looping while avoiding making the query in python by calling methods.
In the end, the result would be the same, so it's more a matter of taste. (and I had nothing to do with analytics so I'm not sure how you worked with William? I think that reviewed it)

Reply:
Indeed that would be easier, but the goal with taking the query with methods is to get the domain that could come with other rules (like company).

Reply:
Now, I can still indeed do the group by in the sql query instead of doing the count afterwards

Reply:
Indeed I didn't include the company, I wasn't sure if it was necessary in this context as we're working directly with the id's of the accounts (no idea if you can set the same account in different companies, I guess not)

Anyway, it was more of a random idea than an "it's better", the current solution seems ok to me :)